### PR TITLE
[CARBONDATA-2384] SDK support write/read data into/from S3

### PR DESCRIPTION
**Pull request by xubo245**
SDK support write data into S3
```
  val writer = CarbonWriter.builder()
        .withSchema(new Schema(fields))
        .outputPath(writerPath)
        .isTransactionalTable(true)
        .uniqueIdentifier(System.currentTimeMillis)
        .setAccessKey(args(0))
        .setSecretKey(args(1))
        .setEndPoint(args(2))
       .persistSchemaFile(true)
       .buildWriterForCSVInput()
```
SDK support write data from S3
```
  val reader = CarbonReader
      .builder(readPath, "_temp")
      .projection(Array[String]("name", "age", "height"))
      .setAccessKey(args(0))
      .setSecretKey(args(1))
      .setEndPoint(args(2)).build
```

It also support different S3 type, which can be set:
` setAccessKey(String key, String value)`, SK and endpoin also like this. 

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed?
  Add some methods
 - [x] Any backward compatibility impacted?
 No
 - [x] Document update required?

 - [x] Testing done
       examples/spark2/src/main/java/org/apache/carbondata/examples/sdk/SDKS3Example.java
       
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
      No

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5383/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5385/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4218/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4521/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4522/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5386/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4523/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4525/

File: examples/spark2/src/main/scala/org/apache/carbondata/sdk/SDkWriteS3Example.scala [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    val fields: Array[Field] = new Array[Field](3)
+    fields(0) = new Field("name", DataTypes.STRING)
+    fields(1) = new Field("age", DataTypes.INT)
+    fields(2) = new Field("height", DataTypes.DOUBLE)
+
+    try {
+      val builder = CarbonWriter.builder()
+        .withSchema(new Schema(fields))
+        .outputPath(writerPath)
+        .isTransactionalTable(true)
```
Comment (sounakr) [inline examples/spark2/src/main/scala/org/apache/carbondata/sdk/SDkWriteS3Example.scala]: Please write test cases for Non Transactional table in reader and writer.

Comment (xuchuanyin) [reply examples/spark2/src/main/scala/org/apache/carbondata/sdk/SDkWriteS3Example.scala]: @sounakr @xubo245
 What does the meaning of `Transactional table` in carbondata?

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5408/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4546/

File: examples/spark2/src/main/scala/org/apache/carbondata/sdk/SDkWriteS3Example.scala (hunk 1 above)
Comment (xubo245) [reply examples/spark2/src/main/scala/org/apache/carbondata/sdk/SDkWriteS3Example.scala]: @sounakr I change the code to support configure write Non Transactional table. But not carbonReader don't support read Non Transactional table. I will raise another PR to support this function.

Comment (xubo245) [reply examples/spark2/src/main/scala/org/apache/carbondata/sdk/SDkWriteS3Example.scala]: If  isTransactionalTable set false, writes the carbondata and carbonindex files in a flat folder structure. Please check https://github.com/apache/carbondata/blob/master/docs/sdk-writer-guide.md. @xuchuanyin

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4262/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5429/

**Comment by xubo245** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4284/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5446/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4568/

**Comment by xubo245** [comment]
@jackylk @sounakr CI pass. please check it again.

**Comment by jackylk** [comment]
LGTM

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4572/

**Pull request closed without merging**